homeassistant/components/amcrest/__init__.py

```diff
@@ -222,68 +102,126 @@ def _start_event_monitor(hass, name, api, event_codes):
+
+    # currently aiohttp only works with basic authentication
+    # only valid for mjpeg streaming
+    if (
+        config_entry.options.get(CONF_AUTHENTICATION, DEFAULT_AUTHENTICATION)
+        == HTTP_BASIC_AUTHENTICATION
+    ):
+        authentication = aiohttp.BasicAuth(username, password)
+    else:
+        authentication = None
 
-        ffmpeg_arguments = device[CONF_FFMPEG_ARGUMENTS]
-        resolution = RESOLUTION_LIST[device[CONF_RESOLUTION]]
-        binary_sensors = device.get(CONF_BINARY_SENSORS)
-        sensors = device.get(CONF_SENSORS)
-        stream_source = device[CONF_STREAM_SOURCE]
-        control_light = device.get(CONF_CONTROL_LIGHT)
-
-        # currently aiohttp only works with basic authentication
-        # only valid for mjpeg streaming
-        if device[CONF_AUTHENTICATION] == HTTP_BASIC_AUTHENTICATION:
-            authentication = aiohttp.BasicAuth(username, password)
-        else:
-            authentication = None
-
-        hass.data[DATA_AMCREST][DEVICES][name] = AmcrestDevice(
-            api,
-            authentication,
-            ffmpeg_arguments,
-            stream_source,
-            resolution,
-            control_light,
+    hass.data[DATA_AMCREST][DEVICES][name] = AmcrestDevice(
+        api,
+        authentication,
+        ffmpeg_arguments,
+        stream_source,
+        resolution,
+        control_light,
+    )
+
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(config_entry, CAMERA)
+    )
+
+    event_codes = dict()
+    event_codes["binary_sensors"] = []
+    event_codes["events"] = []
+
+    if binary_sensors:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, BINARY_SENSOR)
         )
 
-        discovery.load_platform(hass, CAMERA, DOMAIN, {CONF_NAME: name}, config)
+        event_codes["binary_sensors"] = [
+            BINARY_SENSORS[sensor_type][SENSOR_EVENT_CODE]
+            for sensor_type in binary_sensors
+            if sensor_type not in BINARY_POLLED_SENSORS
+        ]
 
-        if binary_sensors:
-            discovery.load_platform(
-                hass,
-                BINARY_SENSOR,
-                DOMAIN,
-                {CONF_NAME: name, CONF_BINARY_SENSORS: binary_sensors},
-                config,
-            )
-            event_codes = [
-                BINARY_SENSORS[sensor_type][SENSOR_EVENT_CODE]
-                for sensor_type in binary_sensors
-                if sensor_type not in BINARY_POLLED_SENSORS
-            ]
-            if event_codes:
-                _start_event_monitor(hass, name, api, event_codes)
+    if events:
+        event_codes["events"] = [event.strip() for event in events.split(",")]
 
-        if sensors:
-            discovery.load_platform(
-                hass, SENSOR, DOMAIN, {CONF_NAME: name, CONF_SENSORS: sensors}, config
-            )
+    if event_codes["binary_sensors"] or event_codes["events"]:
```

Comment:
if it's not a problem I prefer my code, it's more readable in my opinion

Reply:
My code was completely wrong anyway: `any(codes for codes in event_codes)` is a convoluted way to say `any(event_codes)` which in fact means `any(event_codes.keys())`. And it would be always true since the keys themselves are non-empty strings...

The correct code is `if any(event_codes.values()):` which is arguably much simpler and more generic (meaning « if there is any non-nul event code category »).